### ISSUE TITLE
Some texture glitches with new PNG Decoder

### ISSUE THREAD

**Issue by Kenny-Hui** [leezer3/OpenBVE#943]
### Description

On the latest nightly build with the new PNG Decoder, certain textures appears to be glitched.
![image](https://github.com/leezer3/OpenBVE/assets/28094366/9eae6bcf-c07d-4bed-bc8c-50f382705eff)
![image](https://github.com/leezer3/OpenBVE/assets/28094366/7db13948-3c13-43c8-bc58-3c665939a5ba)

I had encountered similar patterns before with something like MSI Afterburner Overlay (#566), but there should be no external program manipulating the renderer and I can reproduce this consistently, as well as under Wine.

### Reproduction

If the issue occurs in multiple routes/ trains, please provide one or two samples.
In order to reproduce the issue and debug it, it's helpful to have the following:

## Route

[MOL.zip](https://github.com/leezer3/OpenBVE/files/9373760/MOL.zip) from #807 

### Related information 

* Operating system: Debian Testing (Trixie)
* Method of control: Keyboard

**Comment by leezer3** [comment]
Now that's wierd.
(Full train under Windows)
![test](https://github.com/leezer3/OpenBVE/assets/7688245/e951b61b-9e15-4759-82c7-e1f890fce09f)

(Handle under Debian, no alpha)
![test2](https://github.com/leezer3/OpenBVE/assets/7688245/26ff4a6e-dff0-4a36-8ab3-4f41e0978bee)

Your handles and other textures definitely work OK here.
The textures causing glitches appear to be RGB (as opposed to RGBA), not respecting the power of 2 rule.

---

The basic difference would appear to be that the new PNG decoder uploads to the graphics card as RGB where possible, whereas the original uploads as RGBA. 

What graphics card is this?
I would suspect a driver bug somewhere in the first instance. 
It's also *really* odd that it's corrupting the RGB pixels, but not the alpha on the handle. (I'd expect the whole thing to be a glitched mess if the decoder was interpreting something incorrectly)

**Comment by Kenny-Hui** [comment]
Sorry that was an older variants of the train, the new version I have uses a 3D cylinder along with a dedicated handrail texture, so no alpha should be involved. (Although it still somehow does not conform to the Po^2 rules, that should be an easy fix but it seems to affect other legacy content as well)

The newer train version is in [here](https://github.com/HKTSS/LRV-P4/tree/v2.5).
The **MOL.zip** route attached to this issue should also be able to reproduce, though you need to travel to the 2nd station to see another track variant with the broken texture.

The GPU used here is an NVIDIA GTX 1660 SUPER

**Comment by leezer3** [comment]
![test3](https://github.com/leezer3/OpenBVE/assets/7688245/66719bc5-604f-4b88-a527-0ffa9cd28f44)

![test4](https://github.com/leezer3/OpenBVE/assets/7688245/113fa541-6c5f-4bea-9023-29901e579639)

No corruption I'm afraid in either the VM or Windows. I'll try native Debian in a minute. 
yl_handrail.png is RGBA, which is marginally interesting (the alpha channel isn't of any use)

Is this using Nouveau or the Nvidia proprietary drivers?

**Comment by leezer3** [comment]
Interesting:
I can now reproduce the MOL glitch, but only in Route Viewer and the main game, but not Object Viewer.
This isn't actually a PNG either- it's a BMP.

I can't however reproduce your issues with the handrail texture on native Debian Bookworm, KDE plasma & latest proprietary Nvidia drivers.

**Comment by leezer3** [comment]
OK, I think I understand what's happening now.

Please test the most recent nightly.
TLDR:
This was order dependant, so whilst I can't reproduce your handle gitch at the minute, I think it should sort out.

**Comment by Kenny-Hui** [comment]
Yup that fixes it
![image](https://github.com/leezer3/OpenBVE/assets/28094366/5ee7f35a-4616-41a5-855e-5887935eb3ca)

Also I've noticed some very weird straight line pattern going on with the seat which doesn't seems exist on the version before.
Seems like some bits gets shifted to the top, although I have not touched the seat code for quite a while now (It's a bit of a mess IIRC) so I can't tell what is happening, maybe I'll have a look at it later when I have more time.
![image](https://github.com/leezer3/OpenBVE/assets/28094366/d72826c0-65e9-4104-8d0c-ece61b02021a)

**Comment by leezer3** [comment]
Not sure that's related.
Looks to me like texture bleeding. You might want to double-check the quality settings.

**Comment by leezer3** [comment]
Now this is seriously interesting.....

Convert the texture to BMP, and the bleeding goes away.
Similarly, re-save the PNG in paint / Irfanview or compress with PNGGauntlet and it's OK.

Let me dig a little deeper, there must be something a little odd about this one.

**Comment by leezer3** [comment]
Hit *again*, might work hopefully this time.....

Michelle's original code expected textures to internally always be loaded as RGBA. 
There's some interesting back-blending to handle this case, which actually seems to cause this when the texture is *not* RGBA. (I'm presuming her original code is for textures using an alpha channel *and* a transparent statement, stupid as this is.....)

**Comment by Kenny-Hui** [comment]
Thanks, *should* be fully resolved.
